relations/jf/src/note.rs

```diff
@@ -0,0 +1,167 @@
+    CircuitField, PlonkResult, PublicInput,
+};
+
+pub enum NoteType {
+    Deposit,
+    Spend,
+}
+
+pub struct SourcedNote {
+    pub note: Note,
+    pub token_id: TokenId,
+    pub token_amount: TokenAmount,
+    pub trapdoor: Trapdoor,
+    pub nullifier: Nullifier,
+    pub note_type: NoteType,
+}
+
+pub struct SourcedNoteVar {
+    pub note_var: Variable,
+    pub token_id_var: Variable,
+    pub token_amount_var: Variable,
+    pub trapdoor_var: Variable,
+    pub nullifier_var: Variable,
+}
+
+pub trait NoteGadget {
+    fn create_note_variable(&mut self, note: &SourcedNote) -> PlonkResult<SourcedNoteVar>;
+    fn enforce_note_preimage(&mut self, note_var: SourcedNoteVar) -> PlonkResult<Variable>;
+}
+
+impl NoteGadget for PlonkCircuit<CircuitField> {
+    fn create_note_variable(&mut self, note: &SourcedNote) -> PlonkResult<SourcedNoteVar> {
+        let note_var = self.create_variable(convert_hash(note.note))?;
+        let token_id_var = self.create_variable(note.token_id.into())?;
+        let token_amount_var = self.create_variable(note.token_amount.into())?;
+        let nullifier_var = self.create_variable(convert_hash(note.nullifier))?;
+        let trapdoor_var = self.create_variable(convert_hash(note.trapdoor))?;
+
+        match note.note_type {
+            NoteType::Spend => {
+                self.set_variable_public(nullifier_var)?;
+            }
+            NoteType::Deposit => {
+                self.set_variable_public(note_var)?;
+                self.set_variable_public(token_id_var)?;
+                self.set_variable_public(token_amount_var)?;
+            }
+        }
+
+        // Ensure that the token amount is valid.
+        // todo: extract token amount limiting to at least constant, or even better to a function/type
+        self.enforce_leq_constant(token_amount_var, CircuitField::from(u128::MAX))?;
+
+        Ok(SourcedNoteVar {
+            note_var,
+            token_id_var,
+            token_amount_var,
+            nullifier_var,
+            trapdoor_var,
+        })
+    }
+
+    fn enforce_note_preimage(&mut self, note_var: SourcedNoteVar) -> PlonkResult<Variable> {
+        let SourcedNoteVar {
+            note_var,
+            token_id_var,
+            token_amount_var,
+            nullifier_var,
+            trapdoor_var,
+        } = note_var;
+
+        let zero_var = self.create_constant_variable(CircuitField::zero())?;
+
+        // Check that the note is valid.
+        let inputs: [usize; 6] = [
+            token_id_var,
+            token_amount_var,
+            trapdoor_var,
+            nullifier_var,
+            zero_var,
+            zero_var,
+        ];
+
+        let computed_note_var = RescueNativeGadget::<CircuitField>::rescue_sponge_no_padding(
+            self,
+            inputs.as_slice(),
+            1,
+        )?[0];
+
+        self.enforce_equal(note_var, computed_note_var)?;
+
+        Ok(token_amount_var)
```

Comment:
we don't have to return this, because the caller already has `note_var: SourcedNoteVar`

Reply:
right, changed.

Reply:
They don't b/c it's passed into the method with ownership.

Reply:
it's `Copy`

Reply:
It _can_ be? https://github.com/Cardinal-Cryptography/aleph-node/pull/1181/files#diff-843190817ebe94d180908160036f30f6275895c0ddd5dd3a315ef8cf090c7ed9R24

Reply:
apologies; should be :) @mike1729

Reply:
Variable is copy, I'm just writing withdraw rel and it's fine :)